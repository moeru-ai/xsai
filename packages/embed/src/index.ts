export { embed as default, embed, type EmbedOptions, type EmbedResponseUsage, type EmbedResult } from './utils/embed'
export { embedMany, type EmbedManyOptions, type EmbedManyResult } from './utils/embed-many'
