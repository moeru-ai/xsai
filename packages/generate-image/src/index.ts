import type { CommonRequestOptions } from '@xsai/shared'

import { requestBody, requestHeaders, requestURL, responseCatch, responseJSON } from '@xsai/shared'

export interface GenerateImageOptions extends CommonRequestOptions {
  [key: string]: unknown
  /**
   * The number of images to generate.
   * @default `1`
   */
  n?: number
  /** A text description of the desired image(s). */
  prompt: string
  /**
   * The format in which generated images with `dall-e-2` and `dall-e-3` are returned from the API.
   * `gpt-image-1` only supports `b64_json`, and `url` will be ignored.
   *
   * Images will always be Base64-encoded in the result. If `url` is used, the images will be
   * fetched upon receiving the response.
   */
  responseFormat?: 'b64_json' | 'url'
  /**
   * The size of the generated images.
   * @default `1024x1024`
   */
  size?: `${number}x${number}`
}

export interface GenerateImageResponse {
  created: number
  data: {
    /**
     * The Bse64-encoded JSON of the generated image. Default value for `gpt-image-1`, and only
     * present if `response_format` is set to `b64_json` for `dall-e-2` and `dall-e-3`.
     */
    b64_json?: string
    /**
     * For `dall-e-3` only, the revised prompt that was used to generate the image.
     */
    revised_prompt?: string
    /**
     * When using `dall-e-2` or `dall-e-3`, the URL of the generated image if `response_format` is
     * set to `url` (default value). Unsupported for `gpt-image-1`.
     */
    url?: string
  }[]
}

export interface GenerateImageResult {
  image: GenerateImageResultImage
  images: GenerateImageResultImage[]
}

export interface GenerateImageResultImage {
  base64: string
  mimeType: string
}

/** @internal */
const mimeTypes = {
  '/9j/': 'image/jpg',
  'AAAAIGZ0eXBhdmlm': 'image/avif',
  'iVBORw0KGgo': 'image/png',
  'R0lGOD': 'image/gif',
  'UklGRg==': 'image/webp',
}

/** @internal */
const convertImage = (b64_json: string) => {
  const key = Object.keys(mimeTypes).find(prefix => b64_json.startsWith(prefix)) as keyof typeof mimeTypes | undefined
  const mimeType = mimeTypes[key ?? 'iVBORw0KGgo']

  return {
    // eslint-disable-next-line @masknet/string-no-data-url
    base64: `data:${mimeType};base64,${b64_json}`,
    mimeType,
  }
}

/** @internal */
const responseBlobAsDataURL = async (res: Response): Promise<string> =>
  responseCatch(res)
    .then(async (res) => {
      const blob = await res.blob()

      try {
        return await new Promise<string>((resolve, reject) => {
          const reader = new FileReader()
          reader.onloadend = () => resolve(reader.result as string)
          reader.onerror = reject
          reader.readAsDataURL(blob)
        })
      }
      catch {
        throw new Error(`Failed to parse response blob, response URL: ${res.url}`)
      }
    })

/** @experimental */
export const generateImage = async (options: GenerateImageOptions): Promise<GenerateImageResult> =>
  (options.fetch ?? globalThis.fetch)(requestURL('images/generations', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateImageResponse>)
    .then(async ({ data }) =>
      Promise.all(
        data.map(async (img, i) => {
          if (typeof img.b64_json === 'string') {
            return convertImage(img.b64_json)
          }

          if (typeof img.url === 'string') {
            return (options.fetch ?? globalThis.fetch)(new URL(img.url), {
              signal: options.abortSignal,
            })
              .then(responseBlobAsDataURL)
              .then((dataURL) => {
                const sepIndex = dataURL.indexOf(';')
                const mimeType = dataURL.substring(5, sepIndex) // `data:(...);...`
                return {
                  base64: dataURL,
                  mimeType,
                }
              })
          }

          throw new Error(`Unrecognized image at index ${i}: ${JSON.stringify(img)}`)
        }),
      ),
    )
    .then(images => ({ image: images[0], images }))
