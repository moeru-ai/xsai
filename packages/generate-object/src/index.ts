import { type Infer, type Schema, toJSONSchema, validate } from '@typeschema/main'
import { generateText, type GenerateTextOptions, type GenerateTextResult } from '@xsai/generate-text'
import { clean } from '@xsai/shared'

export interface GenerateObjectOptions<T extends Schema> extends GenerateTextOptions {
  schema: T
  schemaDescription?: string
  schemaName?: string
}

// TODO: toolCalls, toolResults
export interface GenerateObjectResult<T extends Schema> extends Omit<GenerateTextResult, 'text' | 'toolCalls' | 'toolResults'> {
  object: Infer<T>
}

/** @experimental WIP */
export const generateObject = async <T extends Schema>(options: GenerateObjectOptions<T>): Promise<GenerateObjectResult<T>> =>
  await generateText({
    ...options,
    response_format: {
      json_schema: {
        description: options.schemaDescription,
        name: options.schemaName ?? 'json_schema',
        schema: await toJSONSchema(options.schema)
          .then(json => clean({
            ...json,
            $schema: undefined,
          })),
        strict: true,
      },
      type: 'json_schema',
    },
    schema: undefined,
    schemaDescription: undefined,
    schemaName: undefined,
  })
    .then(async ({ finishReason, steps, text, usage }) => {
      const result = await validate(options.schema, JSON.parse(text!))

      if (result.success) {
        return {
          finishReason,
          object: result.data,
          steps,
          usage,
        }
      }
      else {
        throw new Error([
          'Schema validation failed:',
          ...result.issues.map(issue => `- ${issue.message}`),
        ].join('\n'))
      }
    })

export default generateObject
