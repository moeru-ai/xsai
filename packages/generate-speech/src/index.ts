import { type CommonRequestOptions, requestBody, requestHeaders, requestURL, responseCatch } from '@xsai/shared'

export interface GenerateSpeechOptions extends CommonRequestOptions {
  [key: string]: unknown
  input: string
  /** @default `mp3` */
  responseFormat?: 'aac' | 'flac' | 'mp3' | 'opus' | 'pcm' | 'wav'
  /** @default `1.0` */
  speed?: number
  voice: string
}

/**
 * @experimental WIP
 * @param options generate speech options.
 * @returns audio array buffer
 */
export const generateSpeech = async (options: GenerateSpeechOptions): Promise<ArrayBuffer> =>
  await (options.fetch ?? globalThis.fetch)(requestURL('audio/speech', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseCatch)
    .then(res => res.arrayBuffer())

export default generateSpeech
