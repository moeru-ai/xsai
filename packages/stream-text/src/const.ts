import type { AssistantMessage, FinishReason, Message, ToolCall, Usage } from '@xsai/shared-chat'

// eslint-disable-next-line @masknet/string-no-data-url
export const chunkHeaderPrefix = 'data:'

export interface StreamTextChoice {
  finish_reason?: FinishReason | null
  index: number
  message: StreamTextMessage
}

export interface StreamTextChoiceState {
  calledToolCallIDs: Set<string>
  currentToolID: null | string
  endedToolCallIDs: Set<string>
  index: number
  toolCallErrors: { [id: string]: Error }
  toolCallResults: { [id: string]: string }
}

export interface StreamTextChunkResult {
  choices: {
    delta: {
      content?: string
      refusal?: string
      role: 'assistant'
      tool_calls?: ToolCall[]
    }
    finish_reason?: FinishReason
    index: number
  }[]
  created: number
  id: string
  model: string
  object: 'chat.completion.chunk'
  system_fingerprint: string
  usage?: Usage
}

export interface StreamTextMessage extends Omit<AssistantMessage, 'tool_calls'> {
  content?: string
  tool_calls?: { [id: string]: StreamTextToolCall }
}

export interface StreamTextStep {
  choices: StreamTextChoice[]
  messages: Message[]
  usage?: Usage
}

export interface StreamTextToolCall extends ToolCall {
  function: {
    arguments: string
    name: string
    parsed_arguments: Record<string, unknown>
  }
}
