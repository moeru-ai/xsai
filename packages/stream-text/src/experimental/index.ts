import type { ChatOptions, CompletionStep, CompletionToolCall, CompletionToolResult, FinishReason, Message, ToolCall, Usage } from '@xsai/shared-chat'

import { objCamelToSnake, trampoline } from '@xsai/shared'
import { chat, determineStepType, executeTool } from '@xsai/shared-chat'

import type { StreamTextEvent } from './types/event'

import { transformChunk } from './_transform-chunk'
import { DelayedPromise } from './internal/delayed-promise'

export interface StreamTextOptions extends ChatOptions {
  /** @default 1 */
  maxSteps?: number
  onEvent?: (event: StreamTextEvent) => Promise<unknown> | unknown
  onFinish?: (step?: CompletionStep) => Promise<unknown> | unknown
  onStepFinish?: (step: CompletionStep) => Promise<unknown> | unknown
  /**
   * If you want to disable stream, use `@xsai/generate-{text,object}`.
   */
  stream?: never
  streamOptions?: {
    /**
     * Return usage.
     * @default `undefined`
     */
    includeUsage?: boolean
  }
}

export interface StreamTextResult {
  fullStream: ReadableStream<StreamTextEvent>
  messages: Promise<Message[]>
  steps: Promise<CompletionStep[]>
  textStream: ReadableStream<string>
  usage: Promise<undefined | Usage>
  // TODO: totalUsage
}

export const streamText = async (options: StreamTextOptions): Promise<StreamTextResult> => {
  // state
  const steps: CompletionStep[] = []
  const messages: Message[] = structuredClone(options.messages)
  const maxSteps = options.maxSteps ?? 1
  let usage: undefined | Usage

  // result state
  const resultSteps = new DelayedPromise<CompletionStep[]>()
  const resultMessages = new DelayedPromise<Message[]>()
  const resultUsage = new DelayedPromise<undefined | Usage>()

  // output
  let eventCtrl: ReadableStreamDefaultController<StreamTextEvent> | undefined
  let textCtrl: ReadableStreamDefaultController<string> | undefined
  const eventStream = new ReadableStream<StreamTextEvent>({ start: controller => eventCtrl = controller })
  const textStream = new ReadableStream<string>({ start: controller => textCtrl = controller })

  const pushEvent = (stepEvent: StreamTextEvent) => {
    eventCtrl?.enqueue(stepEvent)
    // eslint-disable-next-line sonarjs/void-use
    void options.onEvent?.(stepEvent)
  }

  const pushStep = (step: CompletionStep) => {
    steps.push(step)
    // eslint-disable-next-line sonarjs/void-use
    void options.onStepFinish?.(step)
  }

  const startStream = async () => {
    // let stepUsage: undefined | Usage
    const pushUsage = (u: Usage) => {
      usage = u
      // stepUsage = u
    }

    let text: string = ''
    const pushText = (content?: string) => {
      textCtrl?.enqueue(content)
      text += content
    }

    const toolCalls: ToolCall[] = []
    const toolResults: CompletionToolResult[] = []
    let finishReason: FinishReason = 'other'

    await chat({
      ...options,
      maxSteps: undefined,
      messages,
      stream: true,
      streamOptions: options.streamOptions != null
        ? objCamelToSnake(options.streamOptions)
        : undefined,
    }).then(async res => res.body!
      .pipeThrough(transformChunk())
      .pipeTo(new WritableStream({
        abort: (reason) => {
          eventCtrl?.error(reason)
          textCtrl?.error(reason)
        },
        close: () => {},
        write: (chunk) => {
          if (chunk.usage)
            pushUsage(chunk.usage)

          // skip if no choices
          if (chunk.choices == null || chunk.choices.length === 0)
            return

          const choice = chunk.choices[0]

          if (choice.finish_reason != null)
            finishReason = choice.finish_reason

          if (choice.delta.tool_calls?.length === 0 || choice.delta.tool_calls == null) {
            if (choice.delta.content != null) {
              pushEvent({ text: choice.delta.content, type: 'text-delta' })
              pushText(choice.delta.content)
            }
            else if (choice.delta.refusal != null) {
              pushEvent({ error: choice.delta.refusal, type: 'error' })
            }
            else if (choice.finish_reason != null) {
              pushEvent({ finishReason: choice.finish_reason, type: 'finish', usage })
            }
          }
          else {
            // https://platform.openai.com/docs/guides/function-calling?api-mode=chat&lang=javascript#streaming
            for (const toolCall of choice.delta.tool_calls) {
              const { index } = toolCall

              if (!toolCalls.at(index)) {
                toolCalls[index] = toolCall
                pushEvent({ toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call-streaming-start' })
              }
              else {
                toolCalls[index].function.arguments += toolCall.function.arguments
                pushEvent({ argsTextDelta: toolCall.function.arguments, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call-delta' })
              }
            }
          }
        },
      })),
    )

    if (toolCalls.length !== 0) {
      for (const toolCall of toolCalls) {
        pushEvent({ args: toolCall.function.arguments, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call' })

        const { parsedArgs, result } = await executeTool({
          abortSignal: options.abortSignal,
          messages,
          toolCall,
          tools: options.tools,
        })

        pushEvent({ args: parsedArgs, result, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-result' })

        toolResults.push({ args: parsedArgs, result, toolCallId: toolCall.id, toolName: toolCall.function.name })

        // TODO: executeTool return message, toolCall, toolResult
        messages.push({
          content: result,
          role: 'tool',
          tool_call_id: toolCall.id,
        })
      }
    }
    else {
      messages.push({ content: text, role: 'assistant' })
    }

    pushStep({
      finishReason,
      stepType: determineStepType({ finishReason, maxSteps, stepsLength: steps.length, toolCallsLength: toolCalls.length }),
      text,
      toolCalls: toolCalls.map(toolCall => ({
        args: toolCall.function.arguments,
        toolCallId: toolCall.id,
        toolCallType: toolCall.type,
        toolName: toolCall.function.name,
      }) satisfies CompletionToolCall),
      toolResults,
      usage,
    })

    if (toolCalls.length !== 0 && steps.length < maxSteps)
      return async () => startStream()
  }

  try {
    await trampoline(async () => startStream())
  }
  catch (err) {
    eventCtrl?.error(err)
    textCtrl?.error(err)

    resultSteps.reject(err)
    resultMessages.reject(err)
    resultUsage.reject(err)
  }
  finally {
    eventCtrl?.close()
    textCtrl?.close()

    resultSteps.resolve(steps)
    resultMessages.resolve(messages)
    resultUsage.resolve(usage)

    // eslint-disable-next-line sonarjs/void-use
    void options.onFinish?.(steps.at(-1))
  }

  return {
    fullStream: eventStream,
    messages: resultMessages.promise,
    steps: resultSteps.promise,
    textStream,
    usage: resultUsage.promise,
  }
}
