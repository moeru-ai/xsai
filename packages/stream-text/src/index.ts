import { chatCompletion, type ChatCompletionOptions, type FinishReason } from '@xsai/shared-chat'

export interface StreamTextOptions extends ChatCompletionOptions {
  /** if you want to disable stream, use `@xsai/generate-text` */
  stream?: never
  streamOptions?: {
    /**
     * Return usage.
     * @default `undefined`
     * @remarks Ollama doesn't support this, see {@link https://github.com/ollama/ollama/issues/5200}
     */
    usage?: boolean
  }
}

export interface StreamTextResponseUsage {
  completion_tokens: number
  prompt_tokens: number
  total_tokens: number
}

export interface StreamTextResult {
  chunkStream: ReadableStream<StreamTextResponse>
  finishReason?: FinishReason
  textStream: ReadableStream<string>
  usage?: StreamTextResponseUsage
}

// TODO: improve chunk type
export interface StreamTextResponse {
  choices: {
    delta: {
      content: string
      role: 'assistant'
    }
    finish_reason?: FinishReason
    index: number
  }[]
  created: number
  id: string
  model: string
  object: 'chat.completion.chunk'
  system_fingerprint: string
  usage?: StreamTextResponseUsage
}

const dataHeaderPrefix = 'data: '
const dataErrorPrefix = `{"error":`

/**
 * @experimental WIP, does not support function calling (tools).
 */
export const streamText = async (options: StreamTextOptions): Promise<StreamTextResult> =>
  await chatCompletion({
    ...options,
    stream: true,
  })
    .then((res) => {
      if (!res.body) {
        return Promise.reject(res)
      }

      const decoder = new TextDecoder()

      let finishReason: string | undefined
      let usage: StreamTextResponseUsage | undefined
      let buffer = ''

      const rawChunkStream = res.body.pipeThrough(new TransformStream({
        transform: (chunk, controller) => {
          buffer += decoder.decode(chunk)
          const lines = buffer.split('\n\n')
          buffer = lines.pop() || ''

          for (const line of lines) {
            // Some cases:
            // - Empty chunk
            // - :ROUTER PROCESSING from OpenRouter
            if (!line || !line.startsWith(dataHeaderPrefix)) {
              continue
            }

            if (line.startsWith(dataErrorPrefix)) {
              // About controller error: https://developer.mozilla.org/en-US/docs/Web/API/TransformStreamDefaultController/error
              controller.error(new Error(`Error from server: ${line}`))
              continue
            }

            const lineWithoutPrefix = line.slice(dataHeaderPrefix.length)
            if (lineWithoutPrefix === '[DONE]') {
              controller.terminate()
              continue
            }

            const data: StreamTextResponse = JSON.parse(lineWithoutPrefix)
            controller.enqueue(data)

            if (data.choices[0].finish_reason) {
              finishReason = data.choices[0].finish_reason
            }
            if (data.usage) {
              usage = data.usage
            }
          }
        },
      }))

      const [chunkStream, rawTextStream] = rawChunkStream.tee()

      const textStream = rawTextStream.pipeThrough(new TransformStream({
        transform: (chunk, controller) => controller.enqueue(chunk.choices[0].delta.content),
      }))

      return { chunkStream, finishReason, textStream, usage }
    })

export default streamText
