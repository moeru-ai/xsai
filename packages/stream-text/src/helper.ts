import type { StreamTextChunkResult } from './const'

import { chunkHeaderPrefix } from './const'

export const parseChunk = (text: string): [StreamTextChunkResult | undefined, boolean] | never => {
  if (!text || !text.startsWith(chunkHeaderPrefix))
    return [undefined, false]

  // Extract content after "data:" prefix
  const content = text.slice(chunkHeaderPrefix.length)
  // Remove leading single space if present
  const data = content.startsWith(' ') ? content.slice(1) : content

  // Handle special cases
  if (data === '[DONE]') {
    return [undefined, true]
  }

  // TODO: should we use JSON.parse?
  if (data.startsWith('{') && data.includes('"error":')) {
    throw new Error(`Error from server: ${data}`)
  }

  // Process normal chunk
  const chunk = JSON.parse(data) as StreamTextChunkResult

  return [chunk, false]
}
