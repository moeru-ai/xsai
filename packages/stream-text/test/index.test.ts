import { ollama } from '@xsai/providers'
import { clean } from '@xsai/shared'
import { describe, expect, it } from 'vitest'

import type { ChunkResult } from '../src'

import { streamText } from '../src'

// make TS happy
// https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility
declare global {
  interface ReadableStream<R = any> {
    // eslint-disable-next-line ts/method-signature-style
    [Symbol.asyncIterator](): AsyncIterableIterator<R>
  }
}

describe('@xsai/stream-text', () => {
  it('basic', async () => {
    const { textStream } = await streamText({
      ...ollama.chat('llama3.2'),
      messages: [
        {
          content: 'You are a helpful assistant.',
          role: 'system',
        },
        {
          content: 'This is a test, so please answer \'YES\' and nothing else.',
          role: 'user',
        },
      ],
      // streamOptions: { usage: true },
    })

    const result: string[] = []

    for await (const textPart of textStream) {
      result.push(textPart)
    }

    expect(result.join('')).toStrictEqual('YES')
  })

  it('the-quick-brown-fox', async () => {
    let onChunkCount = 0
    let chunkCount = 0

    const { chunkStream, textStream } = await streamText({
      ...ollama.chat('llama3.2'),
      messages: [
        {
          content: 'You are a helpful assistant.',
          role: 'system',
        },
        {
          content: 'This is a test, so please answer \'The quick brown fox jumps over the lazy dog.\' and nothing else.',
          role: 'user',
        },
      ],
      onChunk: () => { onChunkCount++ },
    })

    const chunk: ChunkResult[] = []
    const text: string[] = []

    for await (const chunkPart of chunkStream) {
      chunk.push(clean({
        ...chunkPart,
        created: undefined,
        id: undefined,
      }))
      chunkCount++
    }

    for await (const textPart of textStream) {
      text.push(textPart)
    }

    expect(text.join('')).toBe('The quick brown fox jumps over the lazy dog.')
    expect(text).toMatchSnapshot()
    expect(chunk).toMatchSnapshot()

    expect(onChunkCount).toMatchSnapshot(chunkCount)
  })
})
