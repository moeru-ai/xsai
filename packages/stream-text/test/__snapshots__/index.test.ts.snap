// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@xsai/stream-text > the-quick-brown-fox 1`] = `
[
  "The",
  " quick",
  " brown",
  " fox",
  " jumps",
  " over",
  " the",
  " lazy",
  " dog",
  ".",
  "",
]
`;

exports[`@xsai/stream-text > the-quick-brown-fox 2`] = `
[
  {
    "choices": [
      {
        "delta": {
          "content": "The",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " quick",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " brown",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " fox",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " jumps",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " over",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " the",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " lazy",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " dog",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": ".",
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
  {
    "choices": [
      {
        "delta": {
          "content": "",
          "role": "assistant",
        },
        "finish_reason": "stop",
        "index": 0,
      },
    ],
    "model": "llama3.2",
    "object": "chat.completion.chunk",
    "system_fingerprint": "fp_ollama",
  },
]
`;

exports[`@xsai/stream-text > the-quick-brown-fox 3`] = `11`;
